DDG4/plugins/Geant4SDActions.cpp

```diff
@@ -27,7 +27,57 @@ namespace dd4hep {
 
     namespace {
       struct Geant4VoidSensitive {};
+
+      /// Common code to handle the creation of a calorimeter hit.
+      template <class HANDLER>
+      void handleCalorimeterHit (VolumeID cell,
+                                 const HitContribution& contrib,
+                                 Geant4HitCollection& coll,
+                                 const HANDLER& h,
+                                 const Geant4Sensitive& sd,
+                                 const Segmentation& segmentation)
+      {
+        typedef Geant4Calorimeter::Hit Hit;
+        Hit* hit = coll.findByKey<Hit>(cell);
+        if ( !hit ) {
+          DDSegmentation::Vector3D pos = segmentation.position(cell);
+          Position global;
+
+          // Convert the position relative to the local readout volume
+          // to a global position.
+          if (!segmentation.cellsSpanVolumes()) {
+            global = h.localToGlobal(pos);
+          }
+          else {
+            // The segmentation can gang together multiple volumes.
+            // In this case, we can't use the transformation we get from
+            // the step --- the volume that actually contains the hit
+            // may not be the same volume that the segmentation uses
+            // for the local coordinate system.  We need to get the
+            // actual volID used from the segmentation and then look
+            // it up the volume manager to get the proper transformation.
+            VolumeID volID = segmentation.volumeID(cell);
+            VolumeManager vman = VolumeManager::getVolumeManager(sd.detectorDescription());
+            VolumeManagerContext* vc = vman.lookupContext(volID);
+            global = vc->localToWorld(Position(pos)) / dd4hep::mm;
```

Comment:
Why is the unit transformation needed here?

Reply:
The localToWorld() call returns something in dd4hep coordinates, but global is used to initialize a Geant4Calorimeter::Hit, which uses Geant4 /CLHEP units (cf Geant4OUtput2EDM4Hep.cpp, which divides the position by CLHEP::mm).

In the existing branch of the code

```
            global = h.localToGlobal(pos);
```

the method used here is

```
dd4hep::Position Geant4HitHandler::localToGlobal(const DDSegmentation::Vector3D& local)  const   {
  return localToGlobal(G4ThreeVector(local.X / dd4hep::mm,local.Y / dd4hep::mm,local.Z / dd4hep::mm));
}
```

which does the unit conversion.

Reply:
Thanks! Let's add a comment for this in the code, please!
```suggestion
            // explicit unit conversion; h.localToGlobal does it internally already
            global = vc->localToWorld(Position(pos)) / dd4hep::mm;
```